NDTensors/src/lib/UnallocatedArrays/src/abstractunallocatedarray.jl

```diff
@@ -0,0 +1,65 @@
+using FillArrays: FillArrays, getindex_value
+using NDTensors.SetParameters: set_parameters
+using Adapt: adapt
+
+const UnallocatedArray{ElT,N,AxesT,AllocT} = Union{
+  UnallocatedFill{ElT,N,AxesT,AllocT},UnallocatedZeros{ElT,N,AxesT,AllocT}
+}
+
+@inline Base.axes(A::UnallocatedArray) = axes(parent(A))
+Base.size(A::UnallocatedArray) = size(parent(A))
+function FillArrays.getindex_value(A::UnallocatedArray)
+  return getindex_value(parent(A))
+end
+
+function Base.complex(A::UnallocatedArray)
+  return complex(eltype(A)).(A)
+end
+
+function Base.transpose(a::UnallocatedArray)
+  return set_alloctype(transpose(parent(a)), alloctype(a))
+end
+
+function Base.adjoint(a::UnallocatedArray)
+  return set_alloctype(adjoint(parent(a)), alloctype(a))
+end
+
+function set_alloctype(T::Type{<:UnallocatedArray}, alloc::Type{<:AbstractArray})
+  return set_parameters(T, Position{4}(), alloc)
+end
+
+## This overloads the definition defined in `FillArrays.jl`
+for STYPE in (:AbstractArray, :AbstractFill)
+  @eval begin
+    @inline $STYPE{T}(F::UnallocatedArray{T}) where {T} = F
+    @inline $STYPE{T,N}(F::UnallocatedArray{T,N}) where {T,N} = F
+  end
+end
+
+function allocate(f::UnallocatedArray)
+  a = similar(f)
+  fill!(a, getindex_value(f))
+  return a
+end
+
+function allocate(arraytype::Type{<:AbstractArray}, elt::Type, axes::Tuple)
+  ## TODO rewrite this using set_eltype and set_ndims functions
+  ## currently these functions are defined in `NDTensors`
+  ## In the future they should be defined in `SetParameters`
+  ArrayT = set_parameters(arraytype, Position{1}(), elt)
+  ArrayT = set_parameters(ArrayT, Position{2}(), length(axes))
+  return similar(ArrayT, axes)
```

Comment:
```suggestion
  arraytype_elt = set_parameters(arraytype, Position{1}(), elt)
  return similar(arraytype_elt, axes)
```
My understanding is that specifying the number of dimensions isn't necessary since it is determined by the length of `axes` anyway.

Reply:
This is correct, I remoed the `set_parameter(..., Position{2}(),,,)` line